compress.go

```diff
@@ -0,0 +1,238 @@
+package mysql
+
+import (
+	"bytes"
+	"compress/zlib"
+	"io"
+)
+
+const (
+	minCompressLength = 50
+)
+
+
+type packetReader interface {
+	readNext(need int) ([]byte, error)
+}
+
+type compressedReader struct {
+	buf      packetReader
+	bytesBuf []byte
+	mc       *mysqlConn
+	zr       io.ReadCloser
+}
+
+type compressedWriter struct {
+	connWriter io.Writer
+	mc         *mysqlConn
+	zw         *zlib.Writer
+}
+
+func NewCompressedReader(buf packetReader, mc *mysqlConn) *compressedReader {
+	return &compressedReader{
+		buf:      buf,
+		bytesBuf: make([]byte, 0),
+		mc:       mc,
+	}
+}
+
+func NewCompressedWriter(connWriter io.Writer, mc *mysqlConn) *compressedWriter {
+	return &compressedWriter{
+		connWriter: connWriter,
+		mc:         mc,
+		zw:         zlib.NewWriter(new(bytes.Buffer)),
```

Comment:
Does it really need to introduce to extra buffers?
Would it make sense to combine the reader and writer (e.g. as a virtual buffer) and share some data, like the `mc` and the buffer?

Reply:
It has been a while since I've looked at my own code so I may be wrong about this, but between `compressedReader` and `compressedWriter`, it appears the only thing they are sharing is a reference to `mc`?  The reader uses a slice, but that slice needs to act as memory between reads so I don't think it would make sense to share it. Please let me know what you meant by this comment!

Reply:
> The reader uses a slice, but that slice needs to act as memory between reads so I don't think it would make sense to share it

That answers my question. Currently we only have one buffer for each connection, which is shared between reads and writes.